Enable mono runtime to handle SIGTERM like CoreCLR #82806

nealef opened the pull request:
Provide fix for #81093 - "Mono does not emit ProcessExit event on SIGTERM"

* src/mono/mono/mini/mini-posix.c
  - Add signal handler for SIGTERM

* src/mono/mono/mini/mini-windows.c
  - Add signal handler for SIGTERM

* src/mono/mono/mini/mini-runtime.c
  - Add mono_sigterm_signal_handler to process SIGTERM that will set a global variable to be monitored by the GC finalizer thread

* src/mono/mono/mini/mini-runtime.h
  - Define prototype for mono_sigterm_signal_handler()

* src/mono/mono/metadata/gc.c
  - Monitor for sigterm and kick off the shutdown process when encountered by calling mono_runtime_try_shutdown().
  - Exit with either the user set exitcode (System.Environment.ExitCode) or SIGTERM + 128.

nealef committed:
Provide fix for #81093 - "Mono does not emit ProcessExit event on SIGTERM"

* src/mono/mono/mini/mini-posix.c
  - Add signal handler for SIGTERM

* src/mono/mono/mini/mini-windows.c
  - Add signal handler for SIGTERM

* src/mono/mono/mini/mini-runtime.c
  - Add mono_sigterm_signal_handler to process SIGTERM that will set a global variable
    to be monitored by the GC finalizer thread

* src/mono/mono/mini/mini-runtime.h
  - Define prototype for mono_sigterm_signal_handler()

* src/mono/mono/metadata/gc.c
  - Monitor for sigterm and kick off the shutdown process when encountered by calling mono_runtime_try_shutdown().
  - Exit with either the user set exitcode (System.Environment.ExitCode) or SIGTERM + 128.

tmds commented:
> What does coreCLR do in this situation? It fields the SIGTERM and invokes managed code. How does it prevent getting stuck?

It doesn't.

The user code may prevent the application from terminating, for example by blocking the `ProcessExit` event.

`SIGTERM` is a friendly request for an app to terminate.

If it doesn't terminate in a timely fashion, the user can send it `SIGKILL`.

That's what a control process, like systemd will do.

@nealef can you remove the following line, which skips the `SigTermExitCode` test on Mono:

https://github.com/dotnet/runtime/blob/db21974134ed636c4e6bbd1a24104dd93c1ebc16/src/libraries/System.Runtime/tests/System/ExitCodeTests.Unix.cs#L19

nealef committed:
* src/libraries/System.Runtime/tests/System/ExitCodeTests.Unix.cs
  - Re-enable ExitCodeTests for mono

tmds commented:
The `SigTermExitCode` test is failing, because `Environment.ExitCode` changes aren't getting picked up.

I think the fix may be something like: call `mono_environment_exitcode_set(128+SIGTERM)` at the start of `mono_sigterm_signal_handler`, and call `exit(mono_environment_exitcode_get()` after `mono_runtime_try_shutdown`.

nealef committed:
* src/mono/mono/mini/mini-runtime.c
  - Set a default exit code before starting the SIGTERM processing

tmds commented on src/mono/mono/mini/mini-runtime.c:
```diff
@@ -3633,7 +3633,9 @@ MONO_SIG_HANDLER_FUNC (, mono_sigterm_signal_handler)
 	term_signaled = TRUE;
 
-	mono_gc_finalize_notify ();
+	mono_environment_exitcode_set(128+SIGTERM);	/* Set default exit code */
```
Can you move this before the `term_signaled` assignment?

tmds commented on src/mono/mono/metadata/gc.c:
```diff
@@ -879,6 +882,17 @@ finalizer_thread (gpointer unused)
+
+		/* Just in case we've received a SIGTERM */
+		if (term_signaled) {
+			int ec = mono_environment_exitcode_get();
+			mono_runtime_try_shutdown();
+			if (ec == 0)
+				exit(128+SIGTERM);
+			else
+				exit(ec);
```
```suggestion
			mono_runtime_try_shutdown();
			exit(mono_environment_exitcode_get());
```

tmds replied:
We no longer need the `if (ec == 0)` branch.

nealef committed:
* src/mono/mono/mini/mini-runtime.c
  - Set a default exit code before setting the term_signaled variable that gets checked in gc

nealef committed:
* src/mono/mono/metadata/gc.c
  - Simplify use of exit code now that a default is being set

nealef commented:
Fixed

tmds commented:
Thanks.

Let's see if the test passes for all cases with these changes.

tmds commented:
The test is passing for all cases. So it has the right behavior now concerning the ExitCode.

I'm not familiar enough with mono to approve the implementation.

lambdageek commented on src/mono/mono/mini/mini-runtime.c:
```diff
@@ -118,6 +118,7 @@ const char *mono_build_date;
 gboolean mono_do_signal_chaining;
 gboolean mono_do_crash_chaining;
 int mini_verbose = 0;
+gboolean term_signaled = FALSE;
```
nit: `mono_term_signaled`.
nit: `volatile` ?

nealef replied:
Changed the variable name. As long as it's not spinning on that value I assume it will reload on each reference?

lambdageek commented on src/mono/mono/mini/mini-windows.c:
```diff
@@ -189,6 +189,7 @@ mono_runtime_install_handlers (void)
 	win32_seh_set_handler(SIGFPE, mono_sigfpe_signal_handler);
 	win32_seh_set_handler(SIGILL, mono_crashing_signal_handler);
 	win32_seh_set_handler(SIGSEGV, mono_sigsegv_signal_handler);
+	win32_seh_set_handler(SIGSEGV, mono_sigterm_signal_handler);
```
SIGTERM ?

nealef replied:
D'Oh! What happens when you don't have a Windows environment to build this stuff. Changed.

lambdageek commented on src/mono/mono/metadata/gc.c:
```diff
@@ -879,6 +882,13 @@ finalizer_thread (gpointer unused)
 
 	while (!finished) {
+
+		/* Just in case we've received a SIGTERM */
+		if (term_signaled) {
+			mono_runtime_try_shutdown();
+			exit(mono_environment_exitcode_get());
+		}
+
```
Not sure if it matters since it's a loop, but maybe move this check after the `finalizer_sem` wait:

```c
  if (wait) {
    mono_coop_sem_wait (&finalizer_sem, ...);
  }
  wait = TRUE;

  if (term_signaled) {
    ...
  }

  mono_thread_info_set_flags (...);
  ...
```

nealef replied:
Given its the 1st thing to be checked after the thread is given a nudge it probably makes more sense. Though the other finalizer work would be completed before we checked for shutdown. Are there any issues if `mono_run_background_work()` doesn't get done?

lambdageek reviewed (approved):
LGTM.  couple minor nits.

lambdageek commented:
Not sure what's up with `(Build mono minijit Pri0 Runtime Tests Run windows x64 release)` - might be related?  

ping @lateralusX

nealef committed:
* src/mono/mono/metadata/gc.c
  - Rename term_signaled to match mono style
  - Remove volatile attribute
  - Move testing of shutdown until after the sem wait

* src/mono/mono/mini/mini-runtime.c
  - Rename term_signaled to mono_term_signaled

* src/mono/mono/mini/mini-windows.c
  - Use the correct signal for handler

lateralusX commented on src/mono/mono/mini/mini-windows.c:
```diff
@@ -189,6 +189,7 @@ mono_runtime_install_handlers (void)
 	win32_seh_set_handler(SIGFPE, mono_sigfpe_signal_handler);
 	win32_seh_set_handler(SIGILL, mono_crashing_signal_handler);
 	win32_seh_set_handler(SIGSEGV, mono_sigsegv_signal_handler);
+	win32_seh_set_handler(SIGTERM, mono_sigterm_signal_handler);
```
If this is going to work fully on Windows we also need to handle that signal in win32_seh_set_handler as well as react to the exception that will be generated and passed to seh_vectored_exception_handler.

nealef replied:
Adding the `case SIGTERM` and the corresponding `term_handler` variable in `exceptions_[x86|amd64]` appears straightforward but how are they then used? I see there is exception handling for `EXCEPTION_ILLEGAL_INSTRUCTION` etc. where the handler is picked up and called, but how and where would this particular signal be fielded?

nealef replied:
@lateralusX @lambdageek Looking for further guidance on the outstanding Windows' issue.

lateralusX commented on src/mono/mono/metadata/gc.c:
```diff
@@ -892,6 +896,12 @@ finalizer_thread (gpointer unused)
 		wait = TRUE;
 
+		/* Just in case we've received a SIGTERM */
+		if (mono_term_signaled) {
```
Since this is updated from a different thread without any locking, maybe we would need a read memory barrier to make sure we see updated value, especially when wait == FALSE,. There might be one hidden in the mono_coop_sem_timedwait, looks like at least sem_trywait seems to issue memory barriers, and other implementations probably do as well (like WaitForSingleObjectEx), so maybe not needed in the end, but we depend on implementation details I guess we can leave it as is, just wanted to make a comment/note around potential issue.

tmds replied:
I think that at one point, this PR had `mono_term_signaled` as `volatile`.

Maybe that (or something else) is needed to ensure these lines don't reorder, causing the wrong exit code to be picked up.

This is how the exit code is set:

```
mono_environment_exitcode_set(128+SIGTERM);	/* Set default exit code */
mono_term_signaled = TRUE;
```

lateralusX replied:
We can always do acquire/release semantics on mono_term_signaled and we would make sure this is not reordered (volatile will only prevent compiler to reorder load/stores).

nealef commented:
Closing this as a new PR was generated that includes Windows has been created: https://github.com/dotnet/runtime/pull/100056
